### PR TITLE
Refactor validation rules

### PR DESCRIPTION
**Pull request by Shane32**
Replaces #3873 

This allows for post-variable-coercion validation, necessary for the next-gen complexity analyzer (to be done).  It also combines the `IVariableVisitorProvider` into `IValidationRule` which previously was a hidden optional interface layered on top of `IValidationRule`.  It will break any existing `IValidationRule` implementations.

I believe this refactor is long overdue, and due to the other breaking changes around `IMetadataProvider`, might as well be now.

Having all potential validation tie-in points within a single interface will simplify configuration of complex validation rules.  For example, users still just write this in their startup:

```cs
.AddValidationRule<MyValidationRule>()
```

and the validation rule can tie into the GraphQL validation pipeline at all/any necessary points.

### CODE REVIEW

File: src/GraphQL/Validation/Rules.Custom/InputFieldsAndArgumentsOfCorrectLength.cs [hunk 1]
```diff
@@ -18,7 +18,7 @@ namespace GraphQL.Validation.Rules
     /// Please copy the newer <see cref="PatternMatchingDirective"/> and <see cref="PatternMatchingVisitor"/>
     /// when designing new custom directives that can be used to validate input fields and arguments.
     /// </remarks>
-    public class InputFieldsAndArgumentsOfCorrectLength : IValidationRule, IVariableVisitorProvider
+    public class InputFieldsAndArgumentsOfCorrectLength : ValidationRuleBase
```
Comment (Shane32) [inline src/GraphQL/Validation/Rules.Custom/InputFieldsAndArgumentsOfCorrectLength.cs:21]: This is the sample custom validation rule included with GraphQL.NET that uses two tie-in points within the updated `IValidationRule` interface.  Both previously and now, it only requires a single `.AddValidationRule<InputFieldsAndArgumentsOfCorrectLength>()` line of code to install.  The main difference is that previously the `IVariableVisitorProvider` was hidden, in the sense that you'd have to know how GraphQL.NET works to know that it checks if `rule is IVariableVisitorProivder` and make use of its functionality.  Now, anyone that inspects `IValidationRule` can see the method on the interface.

File: docs2/site/docs/migrations/migration8.md [hunk 2]
```diff
@@ -288,6 +288,33 @@ private class Query
...
+}
+```
+
+This allows for a single validation rule to validate AST structure, validate variable values,
+and/or validate coerced field and directive arguments.
+
+To simplify the creation of validation rules, the `ValidationRuleBase` class has been added,
+which implements the `IValidationRule` interface and provides default implementations for
+all three methods.
+
```
Comment (gao-artur) [inline docs2/site/docs/migrations/migration8.md:317]: Are you going to include the comparison with `Validator` and `Parser` delegates mentioned in the previous PR? I think it was important. https://github.com/graphql-dotnet/graphql-dotnet/pull/3873

**Review by gao-artur — COMMENTED** [review]
Looks good!

File: docs2/site/docs/migrations/migration8.md (hunk 2 above)
Comment (Shane32) [reply docs2/site/docs/migrations/migration8.md:321]: I've added the new documentation for creating custom validation rules into the official GraphQL.NET docs (none existed before), referenced from the migration document.  Ok?  It includes various samples you mentioned.

File: docs2/site/docs/getting-started/query-validation.md [hunk 3]
```diff
@@ -16,3 +26,328 @@ await schema.ExecuteAsync(_ =>
...
+Here are some examples:
+
+```csharp
+// for an input object graph type
+
+Field(x => x.FirstName)
+    .Validate(value =>
+    {
+        if (((string)value).Length >= 10)
+            throw new ArgumentException("Length must be less than 10 characters.");
```
[the remark this answers is not in the document]
Comment (Shane32) [reply docs2/site/docs/getting-started/query-validation.md:45]: Yes, and currently no special handling of `ValidationError` or `ExecutionError` exists, so all exceptions are wrapped.  Elsewhere generally throughout GraphQL.NET we allow exceptions that inherit from `ExecutionError` to pass through to the response without being wrapped.  I'm thinking about writing a PR to change this.

File: docs2/site/docs/getting-started/query-validation.md [hunk 4]
```diff
@@ -16,3 +26,328 @@ await schema.ExecuteAsync(_ =>
...
+
+Field(x => x.Password)
+    .Validate(value =>
+    {
+        VerifyPasswordComplexity((string)value);
+    });
+```
+
+The `Validator` delegate is called during the validation stage, prior to execution of the request.
+Null values are not passed to the validation function. Supplying an invalid value will produce
```
[the remark this answers is not in the document]
Comment (Shane32) [reply docs2/site/docs/getting-started/query-validation.md:63]: Yes; that's clearer.

**Review by gao-artur — APPROVED** [review]

**Pull request merged**